### PR TITLE
Fix declaring product of straight-to-jar compilation

### PR DESCRIPTION
**Pull request by adpi2**
Currently the straight-to-JAR compilation does not declare any class file to Zinc, which prevents Zinc from compiling downstream modules incrementally.

For the context, we would like to enable straight-to-JAR compilation in sbt 2.x, and put remote caching on top of it.

### CODE REVIEW

File: compiler/src/dotty/tools/backend/jvm/ClassfileWriters.scala [hunk 1]
```diff
@@ -182,7 +181,7 @@ class ClassfileWriters(frontendAccess: PostProcessorFrontendAccess) {
       jarWriter.putNextEntry(entry)
       try jarWriter.write(bytes, 0, bytes.length)
       finally jarWriter.flush()
-      null
+      new VirtualFile(file.absolutePath + "!" + relativePath, bytes)
```
Comment (adpi2) [inline compiler/src/dotty/tools/backend/jvm/ClassfileWriters.scala:184]: For a file inside a JAR we create a `VirtualFile` with path: `/path/to/file.jar!relative/path/to/file.class `

This is also the convention in Zinc: https://github.com/sbt/zinc/blob/68b9c1a7e203dd20dea40e293574c50f5b22f662/internal/compiler-bridge/src/main/scala/xsbt/JarUtils.scala#L46-L48

Comment (bishabosha) [reply compiler/src/dotty/tools/backend/jvm/ClassfileWriters.scala:184]: So my comment here is that VirtualFile is hardcoded to return null in its `jpath` which is needed by `onFileCreated` - so I think this is the wrong abstraction - perhaps PlainFile with the name funny path

**Comment by adpi2** [comment]
@jamie For a class file in a JAR, I returned the `PlainFile` as you suggested. Also I added a test.

**Comment by bishabosha** [comment]
`xsbt.ExtractUsedNamesSpecification` has started failing - looks like there are classes not found?

**Comment by adpi2** [comment]
> xsbt.ExtractUsedNamesSpecification has started failing - looks like there are classes not found?

Fixed

**Review by bishabosha — APPROVED** [review]

**Comment by adpi2** [comment]
@bishabosha What should I do to get this backported to 3.3.x?

**Comment by sjrd** [comment]
Wait until the backporting bot gets to this PR. There's a lot of backlog but it's basically automated.

**Pull request merged**